storage/zynq7000-flash/zynq7000-flash.c

```diff
@@ -0,0 +1,521 @@
+	printf("\t\tstart:  partition start in bytes\n");
+	printf("\t\tsize:   partition size in bytes\n");
+	printf("\t-r <dev:start:size:fs>       - mount root filesystem\n");
+	printf("\t\tdev:    device name\n");
+	printf("\t\tstart:  partition start in bytes\n");
+	printf("\t\tsize:   partition size in bytes\n");
+	printf("\t\tfs:     filesystem name\n");
+	printf("\t-h                           - print this help message\n");
+}
+
+
+static int flash_oidResolve(const char *devPath, oid_t *oid)
+{
+	char temp[32];
+
+	if (strncmp("/dev/", devPath, 5) != 0) {
+		return -EINVAL;
+	}
+
+	snprintf(temp, sizeof(temp), "devfs/%s", devPath + 5);
+
+	return lookup(temp, NULL, oid);
+}
+
+
+static int flash_createMtdDev(const storage_t *strg, oid_t *oid)
+{
+	int res;
+	char path[32];
+	storage_t *part;
+	unsigned int partID = 0;
+
+	/* Find id for a new partition */
+	if (strg->parent != NULL) {
+		part = strg->parent->parts;
+		if (part != NULL) {
+			do {
+				partID++;
+				part = part->next;
+			} while (part != strg->parent->parts);
+		}
+	}
+
+	/* Add mtdchar device */
+	if (strg->dev->mtd != NULL) {
+		oid->id &= ~MTD_MASK;
+		oid->id |= MTD_CHAR;
+
+		res = snprintf(path, sizeof(path), "/dev/mtd%u", strg->dev->id);
+		if (strg->parent != NULL) {
+			snprintf(path + res, sizeof(path) - res, "p%u", partID);
+		}
+
+		res = create_dev(oid, path);
+		if (res < 0) {
+			fprintf(stderr, "zynq7000-flash: failed to create a device file, err: %d\n", res);
+			return res;
+		}
+	}
+
+	/* Add mtdblock device */
+	if (strg->dev->blk != NULL) {
+		oid->id &= ~MTD_MASK;
+		oid->id |= MTD_BLOCK;
+
+		res = snprintf(path, sizeof(path), "/dev/mtdblock%u", strg->dev->id);
+		if (strg->parent != NULL) {
+			snprintf(path + res, sizeof(path) - res, "p%u", partID);
+		}
+
+		res = create_dev(oid, path);
+		if (res < 0) {
+			fprintf(stderr, "zynq7000-flash: failed to create a device file, err: %d\n", res);
+			return res;
+		}
+	}
+
+	return EOK;
+}
+
+
+static int flash_partAdd(const char *parentPath, off_t start, size_t size)
+{
+	int err;
+	oid_t poid, oid;
+	storage_t *strg, *parent;
+
+	err = flash_oidResolve(parentPath, &poid);
+	if (err < 0) {
+		fprintf(stderr, "zynq7000-flash: cannot resolve %s\n", parentPath);
+		return err;
+	}
+
+	parent = storage_get(GET_STORAGE_ID(poid.id));
+	if (parent == NULL) {
+		err = -EINVAL;
+		fprintf(stderr, "zynq7000-flash: failed to find a parent %s, err: %d\n", parentPath, err);
+		return err;
+	}
+
+	strg = malloc(sizeof(storage_t));
+	if (strg == NULL) {
+		err = -ENOMEM;
+		fprintf(stderr, "zynq7000-flash: failed to allocate a device, err: %d\n", err);
+		return err;
+	}
+
+	strg->start = start;
+	strg->size = size;
+	strg->parent = parent;
+	strg->dev = parent->dev;
+	strg->parts = NULL;
+
+	err = storage_add(strg, &oid);
+	if (err < 0) {
+		free(strg);
+		fprintf(stderr, "zynq7000-flash: failed to create a partition, err: %d\n", err);
+		return err;
+	}
+
+	err = flash_createMtdDev(strg, &oid);
```

Comment:
shouldn't partitions be e.g. /dev/mtdblock0p1 etc instead of /dev/mtdblock1 etc?

Reply:
Take a look on the attached example. Access to the whole device is via file `/dev/mtd0` or `/dev/mtblock0`. Partitions on the `mtd0` device are created adding a suffix `pX` (X - partition's id) - `/dev/mtd0p1`, `/dev/mtdblock0p1`.